src/braket/default_simulator/unitary_matrix_simulation.py

```diff
@@ -0,0 +1,112 @@
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+
+from string import ascii_lowercase, ascii_uppercase
+from typing import List, Tuple
+
+import numpy as np
+import opt_einsum
+
+from braket.default_simulator.operation import GateOperation
+from braket.default_simulator.simulation import Simulation
+
+
+class UnitaryMatrixSimulation(Simulation):
+    """
+    This class tracks the evolution of the unitary matrix of a quantum system with
+    `qubit_count` qubits. The unitary of system evolves by applications of `GateOperation`s
+    using the `evolve()` method.
+    """
+
+    def __init__(self, qubit_count: int, shots: int):
+        """
+        Args:
+            qubit_count (int): The number of qubits being simulated.
+            shots (int): The number of samples to take from the simulation.
+                0 is the only valid value for this simulator.
+        """
+        super().__init__(qubit_count=qubit_count, shots=shots)
+        initial_unitary = np.eye(2 ** qubit_count, dtype=complex)
+        self._unitary_matrix = initial_unitary
+
+    def evolve(self, operations: List[GateOperation]) -> None:
+        """Evolves the unitary of the simulation under the action of
+        the specified gate operations.
+
+        Args:
+            operations (List[GateOperation]): Gate operations to apply for
+                evolving the unitary of the simulation.
+
+        Note:
+            This method mutates the state of the simulation.
+        """
+        self._unitary_matrix = UnitaryMatrixSimulation._apply_operations(
+            self._unitary_matrix, self._qubit_count, operations
+        )
+
+    @staticmethod
+    def _apply_operations(
+        unitary: np.ndarray, qubit_count: int, operations: List[GateOperation]
+    ) -> np.ndarray:
+        un_tensor = np.reshape(unitary, qubit_count * [2, 2])
+        for operation in operations:
+            matrix = operation.matrix
+            targets = operation.targets
+            un_tensor = UnitaryMatrixSimulation._apply_operation(
+                un_tensor, qubit_count, matrix, targets
+            )
+
+        return np.reshape(un_tensor, 2 * [2 ** qubit_count])
+
+    @staticmethod
+    def _apply_operation(
+        un_tensor: np.ndarray, qubit_count: int, matrix: np.ndarray, targets: Tuple[int, ...]
+    ) -> np.ndarray:
+        subscripts = UnitaryMatrixSimulation._einsum_subscripts(targets, qubit_count)
+
+        target_count = len(targets)
+        gate_matrix = np.reshape(matrix, target_count * [2, 2])
+
+        return opt_einsum.contract(subscripts, gate_matrix, un_tensor, dtype=complex, casting="no")
+
+    @staticmethod
+    def _einsum_subscripts(targets: Tuple[int, ...], qubit_count: int) -> str:
```

Comment:
This code is based in this one in Qiskit Terra: https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/providers/basicaer/basicaertools.py#L132

Reply:
If using Qiskit code here, it would be nice to acknowledge that fact.

Reply:
Sure :) Although this code is not a copy of that one in Qiskit so, as I see it, it does not feel under the requirements of Apache license ... but I have to admit I am not sure I fully understand all the details of Apache license, I might be wrong about this. Nonetheless, it is worth to reference the sources, should I add a comment in the code? Some advice on this regard would be appreciated.

Reply:
So, per Apache 2, "derivative works" must contain the copyright as well.  However there is of course gray area.  Perhaps here just cite the original source of the code in question.  However, it is really up to the maintainers of this repo to decide such things.

Reply:
Thank you @nonhermitian for your explanation!!! I will wait then for a comment from the reviewers. Regards